### PR TITLE
Backport #1233 to 2.0

### PR DESCRIPTION
**Pull request by kuenishi**
Merge CV's list-objects prefix optimization + Add key length limit [JIRA: RCS-275 (#1233)]

Conflicts:
    src/riak_cs_s3_response.erl
    src/riak_cs_wm_object.erl
    src/riak_cs_wm_object_upload_part.erl

### CODE REVIEW

**Comment by shino** [comment]
- It seems that the first line of commit message is somewhat misleading.
- xref and dialyzer are failing

File: src/riak_cs_s3_response.erl [hunk 1]
```diff
@@ -158,14 +163,16 @@ error_code(ErrorName) ->
...
-status_code(access_denied) ->  403;
-status_code(reqtime_tooskewed) -> 403;
-status_code(bucket_not_empty) ->  409;
-status_code(bucket_already_exists) -> 409;
-status_code(user_already_exists) -> 409;
-status_code(toomanybuckets) -> 400;
+status_code(invalid_access_key_id)         -> 403;
+status_code(invalid_email_address)         -> 400;
+status_code(access_denied)                 -> 403;
+status_code(copy_source_access_denied)     -> 403;
```
Comment (shino) [inline src/riak_cs_s3_response.erl:169]: This error atom is not used anywhere else?

File: src/riak_cs_s3_response.erl [hunk 2]
```diff
@@ -240,6 +247,10 @@ api_error({Tag, _}=Error, RD, Ctx)
                    Ctx);
 api_error({toomanybuckets, Current, BucketLimit}, RD, Ctx) ->
     toomanybuckets_response(Current, BucketLimit, RD, Ctx);
+api_error({invalid_argument, Name, Value}, RD, Ctx) ->
```
Comment (shino) [inline src/riak_cs_s3_response.erl:250]: This clause is not called.

File: src/riak_cs_wm_object.erl [hunk 3]
```diff
@@ -42,14 +42,23 @@ init(Ctx) ->
...
+            ResponseMod:api_error(Reason, RD, Ctx);
+        {ok, ContextWithKey} ->
+            case riak_cs_wm_utils:has_canned_acl_and_header_grant(RD) of
+                true ->
+                    ResponseMod:api_error(canned_acl_and_header_grant,
                                           RD, ContextWithKey);
-        false ->
-            {false, RD, ContextWithKey}
+                false ->
+                    case riak_cs_copy_object:malformed_request(RD) of
```
Comment (shino) [inline src/riak_cs_wm_object.erl:55]: xref complains about this line.

**Comment by kuenishi** [comment]
After several review and thoughts I think we had better not backport to 2.0. It is easier to resolve conflicting key validation when merging back to 2.1.

**Pull request closed without merging**